File: 23/day3.zig
const std = @import("std");

const Number = struct {
    value: u32,
    start: usize,
    end: usize,
};

const Symbol = struct {
    value: u8,
    col: usize,
    fn validPartNumber(self: Symbol, num: Number) bool {
        for (self.col - 1..self.col + 2) |i| {
            for (num.start..num.end) |j| {
                if (i == j) {
                    return true;
                }
            }
        }
        return false;
    }
};

const Part = struct {
    symbol: Symbol,
    numbers: []const u32,
};

pub fn main() !void {
    var file = try std.fs.cwd().openFile("day3.in", .{});
    defer file.close();

    var buffered = std.io.bufferedReader(file.reader());
    var reader = buffered.reader();

    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    const allocator = gpa.allocator();

    var numberList = std.ArrayList([]const Number).init(allocator);
    defer numberList.deinit();

    var symbolList = std.ArrayList([]const Symbol).init(allocator);
    defer symbolList.deinit();

    var buf: [200]u8 = undefined;
    var line: usize = 0;
    while (try reader.readUntilDelimiterOrEof(&buf, '\n')) |current| {
        var numbers = std.ArrayList(Number).init(allocator);
        defer numbers.deinit();
        var symbols = std.ArrayList(Symbol).init(allocator);
        defer symbols.deinit();
        var inside = false;
        var start: usize = 0;
        // std.debug.print("parsing line {d}:\n{s}\n", .{ line, current });
        for (current, 0..) |c, i| {
            switch (c) {
                '0'...'9' => {
                    if (!inside) {
                        inside = true;
                        start = i;
                    }
                },
                else => {
                    if (inside) {
                        const num = try std.fmt.parseInt(u32, current[start..i], 10);
                        // std.debug.print("found number {d}\n", .{num});
                        try numbers.append(.{ .value = num, .start = start, .end = i });
                        inside = false;
                    }
                    if (c == '.') {
                        continue;
                    }
                    // std.debug.print("found symbol {c}\n", .{c});
                    try symbols.append(.{ .value = c, .col = i });
                },
            }
        }
        if (inside) {
            const num = try std.fmt.parseInt(u32, current[start..], 10);
            // std.debug.print("found number {d}\n", .{num});
            try numbers.append(.{ .value = num, .start = start, .end = current.len });
        }
        try numberList.append(try numbers.toOwnedSlice());
        // std.debug.print("numbers: {any}\n", .{numberList.getLast()});
        try symbolList.append(try symbols.toOwnedSlice());
        // std.debug.print("symbols: {any}\n", .{symbolList.getLast()});
        line += 1;
    }

    var parts = std.ArrayList(Part).init(allocator);
    defer parts.deinit();

    line = 0;
    for (symbolList.items) |symbols| {
        const start = if (line == 0) 0 else line - 1;
        const end = if (line == symbolList.items.len - 1) line + 1 else line + 2;
        // std.debug.print("checking for parts on lines {d} to {d}:\n", .{ start, end });
        for (symbols) |symbol| {
            // std.debug.print("getting numbers for symbol {c} at ({d}, {d})\n", .{ symbol.value, line, symbol.col });
            var valid_numbers = [_]u32{0} ** 6;
            var last: usize = 0;
            for (start..end) |j| {
                for (numberList.items[j]) |number| {
                    if (symbol.validPartNumber(number)) {
                        // std.debug.print("found valid number {d} at ({d}, {d}, {d})\n", .{ number.value, j, number.start, number.end });
                        // try valid_numbers.append(number.value);
                        valid_numbers[last] = number.value;
                        last += 1;
                    }
                }
            }
            try parts.append(.{ .symbol = symbol, .numbers = try allocator.dupe(u32, valid_numbers[0..last]) });
            // std.debug.print("Found part {c} with numbers = {any}\n", .{ symbol.value, valid_numbers });
        }
        line += 1;
    }

    var part_sum: u32 = 0;
    for (parts.items) |part| {
        for (part.numbers) |num| {
            part_sum += num;
        }
    }

    var gear_rs: u32 = 0;
    for (parts.items) |part| {
        // std.debug.print("part {c} has {d} numbers\n", .{ part.symbol.value, part.numbers.len });
        if (part.symbol.value == '*' and part.numbers.len == 2) {
            gear_rs += part.numbers[0] * part.numbers[1];
        }
    }

    var stdout = std.io.getStdOut();
    try stdout.writer().print("Sum of parts: {d}\n", .{part_sum});
    try stdout.writer().print("Gear ratio sum: {d}\n", .{gear_rs});
}

File: 23/day3.py
import re
from dataclasses import dataclass
from pprint import pprint


@dataclass
class Number:
    val: int
    start: int
    end: int


@dataclass
class Symbol:
    var: str
    col: int


@dataclass
class Part:
    symbol: Symbol
    numbers: list[Number]


if __name__ == "__main__":
    numberlist = []
    symbollist = []
    with open("day3.in", "r") as f:
        for line in f.readlines():
            numbers = []
            for match in re.finditer(r"(\d+)", line):
                numbers.append(
                    Number(int(match.group(0)), match.start(0), match.end(0))
                )
            numberlist.append(numbers)

            symbols = []
            for match in re.finditer(r"([^\d\.\n])", line):
                symbols.append(Symbol(match.group(0), match.start(0)))
            symbollist.append(symbols)

    # pprint(numberlist)
    # pprint(symbollist)
    # print("=====")

    parts = []
    for line, symbols in enumerate(symbollist):
        start = 0 if line == 0 else line - 1
        end = line + 1 if line == len(symbollist) - 1 else line + 2
        for symbol in symbols:
            part_numbers = []
            valid_range = range(symbol.col - 1, symbol.col + 2)
            for numbers in numberlist[start:end]:
                for number in numbers:
                    if any(
                        pos in valid_range for pos in range(number.start, number.end)
                    ):
                        # print(f"Found number {number.val} for part {symbol.var}")
                        # print(f"number: ({number.start}, {number.end})")
                        # print(f"symbol: {symbol.col}")
                        part_numbers.append(number)
            parts.append(Part(symbol, part_numbers))
            # print(parts[-1])

    # print("=====")

    part_num_sum = 0
    for part in parts:
        # print(f"Part: {part.symbol.var}")
        ps = sum([number.val for number in part.numbers])
        # print(part.numbers)
        # print(ps)
        part_num_sum += ps

    gear_ratio_sum = 0
    for part in parts:
        if part.symbol.var == "*" and len(part.numbers) == 2:
            gear_ratio_sum += part.numbers[0].val * part.numbers[1].val

    print("Sum of parts", part_num_sum)
    print("Gear ratio sum", gear_ratio_sum)
